ml-prims/src/linalg/eig.h

```diff
@@ -34,39 +36,34 @@ namespace LinAlg {
  * @param eig_vectors: eigenvectors
  * @param eig_vals: eigen values
  * @param cusolverH cusolver handle
- * @param mgr device allocator for temporary buffers during computation
- * @{
+ * @param allocator device allocator for temporary buffers during computation
+ * @param stream cuda stream where to schedule work
  */
 template <typename math_t>
 void eigDC(const math_t *in, int n_rows, int n_cols, math_t *eig_vectors,
            math_t *eig_vals, cusolverDnHandle_t cusolverH,
-           DeviceAllocator &mgr) {
-  cudaStream_t stream;
-  CUSOLVER_CHECK(cusolverDnGetStream(cusolverH, &stream));
-
+           std::shared_ptr<deviceAllocator> allocator, cudaStream_t stream) {
   int lwork;
   CUSOLVER_CHECK(cusolverDnsyevd_bufferSize(cusolverH, CUSOLVER_EIG_MODE_VECTOR,
                                             CUBLAS_FILL_MODE_UPPER, n_rows, in,
                                             n_cols, eig_vals, &lwork));
 
-  math_t *d_work = (math_t *)mgr.alloc(sizeof(math_t) * lwork);
-  int *d_dev_info = (int *)mgr.alloc(sizeof(int));
+  device_buffer<math_t> d_work(allocator, stream, lwork);
+  device_buffer<int> d_dev_info(allocator, stream, 1);
 
   MLCommon::Matrix::copy(in, eig_vectors, n_rows, n_cols);
 
   CUSOLVER_CHECK(cusolverDnsyevd(cusolverH, CUSOLVER_EIG_MODE_VECTOR,
                                  CUBLAS_FILL_MODE_UPPER, n_rows, eig_vectors,
-                                 n_cols, eig_vals, d_work, lwork, d_dev_info));
+                                 n_cols, eig_vals, d_work.data(), lwork,
+                                 d_dev_info.data()));
   CUDA_CHECK(cudaGetLastError());
 
   int dev_info;
-  updateHost(&dev_info, d_dev_info, 1);
+  updateHost(&dev_info, d_dev_info.data(), 1);
```

Comment:
To ensure correct stream ordering this needs to be `updateHostAsync` + `cudaStreamSynchronize` if I do not miss anything.